Add support for Mathematica Not operator (!) in parse_mathematica.

utam-1 opened the pull request:
Fixes a KeyError that occurred when trying to parse boolean string from Mathematica that contains a Not operator : '!'. 

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #27859 

#### Brief description of what is fixed or changed
- Added "Not": Not to_node_conversions in mathematica.py.

- Added tests for basic Not operator parsing, including combinations with And (&&), Or (||), and boolean constants in test_mathematica.py.

- Added special case to handle '!' operator in _parse_after_braces function inside mathematica.py.

-  Added "Factorial": factorial to _node_conversions to ensure postfix factorial (x!) correctly maps to SymPy’s factorial(x) rather than creating a generic Function("Factorial").

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* parsing
  * Fixed `KeyError` when parsing Mathematica `Not` (`!`) operator in `parse_mathematica`.
  * Added proper conversion for postfix factorial (`x!`) to map to SymPy’s `factorial(x)`.
<!-- END RELEASE NOTES -->

utam-1 committed:
parsing.mathematica: Fix KeyError for '!'.

- Added "Not": Not to_node_conversions.

- Added tests for basic Not operator parsing, including combinations with
  And (&&), Or (||), and boolean constants.

- Added special case to handle '!' operator.

utam-1 committed:
Fixed EOF error.

oscarbenjamin commented on sympy/parsing/mathematica.py:
```diff
@@ -1135,8 +1140,10 @@ def converter(expr):
+        "Not": Not,
         "Function": _parse_Function,
+        "Factorial": factorial,
+        "Factorial2": lambda x: factorial(x) * factorial(x - 1),
```
What is factorial2 here doing? Is this tested?

This formula does not look correct. The associated sympy function is called factorial2 and is not equivalent to `factorial(x) * factorial(x - 1)`.

utam-1 committed:
Removed Factorial2 from _node_conversions.

As per the scope of the current PR factorial2 is not needed to be implemented. Sincere apologies for the mistake.

Upabjojr commented:
I'm a bit late, I would recommend also adding a test in `test_parser_mathematica_tokenizer()`, testing that the S-expression equivalent is correctly parsed before converting to a SymPy expression.
